src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// A <see cref="TagHelper"/> that saves the state of Razor components rendered on the page up to that point.
+    /// </summary>
+    [HtmlTargetElement(TagHelperName, TagStructure = TagStructure.WithoutEndTag)]
+    public class PersistComponentStateTagHelper : TagHelper
+    {
+        private const string TagHelperName = "persist-component-state";
+        private const string PersistenceModeName = "persist-mode";
+        private PersistenceMode? _persistenceMode;
+
+        /// <summary>
+        /// Gets or sets the <see cref="Rendering.ViewContext"/> for the current request.
+        /// </summary>
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="PersistenceMode"/> for the state to persist.
+        /// </summary>
+        [HtmlAttributeName(PersistenceModeName)]
+        public PersistenceMode PersistenceMode
+        {
+            get => _persistenceMode ?? throw new InvalidOperationException("Invalid persistence mode.");
+            set => _persistenceMode = value;
+        }
```

Comment:
Is there any way this could be determined automatically based on the rendermode on the `<component>` tag? Would be nice if we could avoid the risk of customers picking the wrong option and getting confused.

Reply:
Maybe we could play around with making the `component` tag a restricted child of this component and see if that helps. 

```razor
<persist-state render-mode="ServerPrerendered">
   <component type="@typeof(App)" />
</persist-state>
```

Reply:
@SteveSandersonMS Yeah, we can probably keep track of the rendering modes inside the view context for this to automagically detect the mode being rendered and also throw if there are mixed server/client components since we can't determine it in that case.

Reply:
@pranavkm I thought of that too, but I don't think we can do something like that since it would prevent preserving the state of multiple components rendered in different files

Reply:
I pivoted this to automatically detect the render mode (with the choice to override) and removed the automatic state persistence on the single component case